tensorflow/c/c_api.cc

```diff
@@ -2399,9 +2405,24 @@ void TF_AddGradients(TF_Graph* g, TF_Output* y, int ny, TF_Output* x, int nx,
 
     const int first_new_node_id = g->graph.num_node_ids();
 
+    const char* child_scope_name = prefix;
+    if (child_scope_name != nullptr) {
+      // The operation should fail if the provided name prefix has already been
+      // used in this graph
+      for (const auto& pair: g->name_map) {
+        const string& name = pair.first;
+        if (name.compare(0, name.find_last_of('/'), prefix) == 0) {
+          status->status =
+              InvalidArgument("Duplicate node name in graph: '", prefix, "'");
```

Comment:
A node with that name may not actually exist, so the error message seems a bit off.
Will the following work instead?

```c++
if (child_scope_name != nullptr) {
  const string conflict = string(prefix) + "/";
  for (const auto& pair: g->name_map) {
    const string& name = pair.first;
    if (name.compare(prefix) == 0 || str_util::StartsWith(name, conflict)) {
      status->status = InvalidArgument("prefix [", prefix, "] conflicts with existing node in the graph named [", name, "]");
      return;
   }
   ...
```

(`str_util::StartsWith` from https://github.com/tensorflow/tensorflow/blob/6ba9573a702ea9e1290517cce4e5f73a14552ad9/tensorflow/core/lib/strings/str_util.h#L146)

Reply:
Mmmh, you are right, I think I missed some cases... for example, do we want to authorise creating gradients nodes under `gradients/sub` if we already created some under `gradients`? I feel that we don't but both algorithms will let it pass.

I'll write down a few more test cases and find something compliant with all of them.

Reply:
Finally thinking of it, my counter-example is wrong and we should support this case. So I picked the algorithm you proposed.